lib/fennec/evaluator/create_permission/request.ex

```diff
@@ -0,0 +1,72 @@
+defmodule Fennec.Evaluator.CreatePermission.Request do
+  @moduledoc false
+
+  import Fennec.Evaluator.Helper
+  alias Jerboa.Format.Body.Attribute
+  alias Jerboa.Params
+  alias Fennec.TURN
+
+  @lifetime 5 * 60 # MUST be 5mins
+
+  @spec service(Params.t, Fennec.client_info, Fennec.UDP.server_opts, TURN.t)
+    :: {Params.t, TURN.t}
+  def service(params, _client, _server, turn_state) do
+    request_status =
+      {:continue, params, %{}}
+      |> maybe(&verify_allocation/3, [turn_state])
+      |> maybe(&verify_xor_peer_address/2, [])
+      |> maybe(&create_permissions/3, [turn_state])
+
+    case request_status do
+      {:error, error_code} ->
+        {%{params | attributes: [error_code]}, turn_state}
+      {:respond, {new_params, new_turn_state}} ->
+        {new_params, new_turn_state}
+    end
+  end
+
+  defp verify_allocation(params, state, turn_state) do
+    case turn_state do
+      %TURN{allocation: %TURN.Allocation{}} ->
+        {:continue, params, state}
+      _ ->
+        {:error, %Attribute.ErrorCode{code: 437}}
+    end
+  end
+
+  defp verify_xor_peer_address(params, state) do
+    # The request MUST have at least one XORPeerAddress
+    case Params.get_attr(params, Attribute.XORPeerAddress) do
+      %Attribute.XORPeerAddress{} ->
+        {:continue, params, state}
+      _ ->
+        {:error, %Attribute.ErrorCode{code: 400}}
+    end
+  end
+
+  defp create_permissions(params, state, turn_state) do
+    expire_at = System.os_time(:seconds) + @lifetime
+    added_pemissions =
+      for peer = %Attribute.XORPeerAddress{} <- Params.get_attrs(params) do
+        %TURN.Permission{
+          address: peer.address,
+          port: peer.port,
```

Comment:
Peer port is irrelevant in permissions context. https://trac.tools.ietf.org/html/rfc5766#section-9.1

```
The port portion of each XOR-PEER-ADDRESS attribute will be ignored and can be any arbitrary value.
```